### DISCUSSION TITLE
왜 groupBy() 를 하고 해야하나?

### DISCUSSION THREAD

**Discussion by madplay** [NMP-Study/spark2023#34] [Q&A]
https://github.com/NMP-Study/spark2023/issues/11 내용 중

> 스트리밍 데이터에 대해 DataFrame.count() 또는 Dataset.reduce() 같은 직접적인 집계 연산은 사용할 수 없다. 이유는 스트리밍 데이터 프레임이 집계 결과를 계속 업데이트 하는 반면, 일반적인 정적 데이터 프레임은 즉시 최종 계산된 결과를 돌려주기 때문이다.
따라서 스트리밍 데이터에 대한 집계에는 항상 DataFrame.groupBy() 또는 Dataset.groupByKey() 함수를 사용해야 한다.

**Comment by KimYunsang-v** [comment]
- 결과값(ex. Long)을 돌려주는 액션 연산을 사용하면 안된다.
- DataFrame을 돌려주는 트랜스포메이션 연산을 사용해야한다.

### groupBy().count() 
```scala
def count(): DataFrame = toDF(Seq(Alias(Count(Literal(1)).toAggregateExpression(), "count")())) 
```

=> DataFrame 반환 

### count() 
```scala
def count(): Long = withAction("count", groupBy().count().queryExecution) { plan =>
  plan.executeCollect().head.getLong(0)
}
```
=> Long 반환
